### PR TITLE
[NFC] Refactor RemoveUsedModuleElements to clarify references and uses

### PR DESCRIPTION
**Pull request by kripken**
This pass talked about reachability and uses and such, but it wasn't very clear
on those things. This refactors it to clarify that we look for references and uses,
and what that means. Specifically, a reference to something makes us keep it
alive, as we need to refer to it; a use makes us also keep it identical in its
contents so that when it is used no behavior changes. A function reference
without a call_ref that can call it is an example of a reference without a use,
which this pass already optimized.

To make that more clear in the code, this refactors out the reference-finding
logic into a new struct, `ReferenceFinder`.

This also replaces the normal walking logic with a more manual traversal
using ChildIterator. This is necessary to properly differentiate references from
uses, but is not immediately useful in this PR except for clarity. It will be
necessary in the next PR, which this prepares for, in which we'll optimize more
reference-but-not-use things.

### CODE REVIEW

File: src/passes/RemoveUnusedModuleElements.cpp [hunk 1]
```diff
@@ -34,36 +48,174 @@
...
 
-  // The set of all reachable things we've seen so far.
-  std::set<ModuleElement> reachable;
+  // Visitors
 
-  // A queue of reachable things that we need to process. These appear in
-  // |reachable|, and the work we do when we pop them from the queue is to look
-  // at the things they reach.
-  std::vector<ModuleElement> queue;
+  void visitCall(Call* curr) {
```
Comment (kripken) [inline src/passes/RemoveUnusedModuleElements.cpp:75]: The code in these visitors is just moved around.

**Comment by kripken** [comment]
Friendly ping.

**Review by tlively — APPROVED** [review]
LGTM with a comment for a suggested follow-up.

File: src/passes/RemoveUnusedModuleElements.cpp [hunk 2]
```diff
@@ -34,36 +48,174 @@
...
-  // signature in here, we know it is reachable.
+  // signature in here, we know it is used; otherwise it may only be referred
+  // to.
   std::unordered_set<HeapType> calledSignatures;
 
   // All the RefFuncs we've seen, grouped by heap type. When we see a CallRef of
   // one of the types here, we know all the RefFuncs corresponding to it are
-  // reachable. This is the reverse side of calledSignatures: for a function to
-  // be reached via a reference, we need the combination of a RefFunc of it as
+  // used. This is the reverse side of calledSignatures: for a function to
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/RemoveUnusedModuleElements.cpp:217]: Makes sense, I'll do a followup as suggested.

**Pull request merged**